### PR TITLE
RATIS-1295. Use a config to enable/disable pre vote

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

Use a config to enable/disable pre vote

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1295

## How was this patch tested?

no need to add new ut

### CODE REVIEW

**Comment by amaliujia** [comment]
Question: shouldn't pre-vote is always a good optimization to accelerate leader election? Is there a situation that it is not suitable thus needs to be disabled?

**Comment by runzhiwang** [comment]
@amaliujia Please see here.
image

**Comment by amaliujia** [comment]
Thanks for explanation! Indeed for larger system it is better to benchmark such changes before adopting it.

File: ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java [hunk 1]
```diff
@@ -555,6 +555,16 @@ static TimeDuration leaderStepDownWaitTime(RaftProperties properties) {
     static void setLeaderStepDownWaitTime(RaftProperties properties, TimeDuration noLeaderTimeout) {
       setTimeDuration(properties::setTimeDuration, LEADER_STEP_DOWN_WAIT_TIME_KEY, noLeaderTimeout);
     }
+
+    String PRE_VOTE_KEY = PREFIX + ".pre-vote";
+    boolean PRE_VOTE_DEFAULT = true;
```
Comment (amaliujia) [inline ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:560]: Nit: might be better to make this default value as `false` so downstream applications (e.g. Ozone) won't feel surprising before benchmarking this optimization.

**Comment by amaliujia** [comment]
Overall LGTM.

We can discuss what could be the default value.

File: ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java (hunk 1 above)
Comment (runzhiwang) [reply ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:560]: If we use false, when we change to true ? I think default true maybe better.

Comment (runzhiwang) [reply ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:560]: @szetszwo What do you think ?

Comment (szetszwo) [reply ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:560]: I do not have a strong opinion.  It seems okay to set true as the default.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

File: ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java (hunk 1 above)
Comment (amaliujia) [reply ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:560]: Sure we can start from `true` then

**Comment by runzhiwang** [comment]
@szetszwo @amaliujia Thanks for review. I have merged the patch.

**Pull request merged**